### ISSUE TITLE
dask.dataframe.sample fail with empty partition, and the fix

### ISSUE THREAD

**Issue by xwang777** [dask/dask#2811]
Hi,

My use case is to sample a random subset from a filtered dataframe, like:

````
from dask import dataframe as dd
ddf = dd.read_parquet(...)
ddf = ddf.query("... some query condition ....")
ddf = ddf.sample(frac=0.1)
df = ddf.compute()
````

It looks like when the query filtering leaves some ddf partitions to be empty, the succeeding sample step will fail with exceptions which basically complains that it cannot sample an empty pandas DataFrame (traceback in the end).

So the fix that I have used is to change the sample function call in dask/dataframe/methods.py to check len(df) before doing sample. Here's the diff between the fixed and current version (0.15.3):

````
166c166
<     return df.sample(random_state=rs, frac=frac, replace=replace) if len(df) > 0 else df
---
>     return df.sample(random_state=rs, frac=frac, replace=replace)
````

Do you think this is a proper fix? If so, can you please apply it to the next version? 

Thanks.

--------------------------------------------------------------------------------------------------------
Traceback of the exception:

distributed.utils - ERROR - a must be greater than 0
Traceback (most recent call last):
  File "/mnt/beegfs/xwang/app/anaconda2/lib/python2.7/site-packages/distributed/utils.py", line 229, in f
    result[0] = yield make_coro()
  File "/mnt/beegfs/xwang/app/anaconda2/lib/python2.7/site-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/mnt/beegfs/xwang/app/anaconda2/lib/python2.7/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "/mnt/beegfs/xwang/app/anaconda2/lib/python2.7/site-packages/tornado/gen.py", line 1063, in run
    yielded = self.gen.throw(*exc_info)
  File "/mnt/beegfs/xwang/app/anaconda2/lib/python2.7/site-packages/distributed/client.py", line 1246, in _gather
    traceback)
  File "/mnt/beegfs/xwang/app/anaconda2/lib/python2.7/site-packages/dask/dataframe/methods.py", line 166, in sample
    return df.sample(random_state=rs, frac=frac, replace=replace)
  File "/mnt/beegfs/xwang/app/anaconda2/lib/python2.7/site-packages/pandas/core/generic.py", line 2900, in sample
    locs = rs.choice(axis_length, size=n, replace=replace, p=weights)
  File "mtrand.pyx", line 1115, in mtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:17104)
ValueError: a must be greater than 0

**Comment by TomAugspurger** [comment]
@xwang777 thanks for the bug report. Yes, your solution sounds reasonable. Are you able to submit a pull request with that fix and a unit test to ensure that it works?

**Comment by xwang777** [comment]
This is the first time I code in github, so it took me quite some time to dig around (and I still cannot git push from my local computer). I finally made a pull request #2818. Please check. Thanks.

**Comment by jcrist** [comment]
Fixed by #2818. Closing.
